File: golang/15.3-sum.go
/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */
package main

import (
	"fmt"
	"sort"
)

// keys
// 1. sort and then use two pointer
// 2. deal with duplication
// 3. choose a good loop condition

// @lc code=start
func threeSum(nums []int) [][]int {
	sort.Sort(sort.IntSlice(nums))

	ret := [][]int{}
	for i := 1; i < len(nums)-1; i++ {

		left := 0
		right := len(nums) - 1

		for left < i && right > i {
			sum := nums[left] + nums[right] + nums[i]
			if sum > 0 {
				right -= 1
				continue
			} else if sum < 0 {
				left += 1
				continue
			}

			duplicated := false
			if left > 0 && nums[left] == nums[left-1] {
				duplicated = true
			}
			if right < len(nums)-1 && nums[right] == nums[right+1] {
				duplicated = true
			}
			if nums[i] == nums[i-1] && left != i-1 {
				duplicated = true
			}
			if !duplicated {
				ret = append(ret, []int{nums[left], nums[i], nums[right]})
			}
			//  it is maybe a good idea to simplely solve situation if sum == target
			// without using continue and break
			if right-1 > i {
				right -= 1
			} else {
				left += 1
			}

		}
	}
	return ret
}

// @lc code=end

func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
}

File: cpp/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 *
 * https://leetcode.com/problems/3sum/description/
 *
 * algorithms
 * Medium (37.57%)
 * Likes:    33838
 * Dislikes: 3162
 * Total Accepted:    5.1M
 * Total Submissions: 13.5M
 * Testcase Example:  '[-1,0,1,2,-1,-4]'
 *
 * Given an integer array nums, return all the triplets [nums[i], nums[j],
 * nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] +
 * nums[k] == 0.
 *
 * Notice that the solution set must not contain duplicate triplets.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [-1,0,1,2,-1,-4]
 * Output: [[-1,-1,2],[-1,0,1]]
 * Explanation:
 * nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
 * nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
 * nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
 * The distinct triplets are [-1,0,1] and [-1,-1,2].
 * Notice that the order of the output and the order of the triplets does not
 * matter.
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [0,1,1]
 * Output: []
 * Explanation: The only possible triplet does not sum up to 0.
 *
 *
 * Example 3:
 *
 *
 * Input: nums = [0,0,0]
 * Output: [[0,0,0]]
 * Explanation: The only possible triplet sums up to 0.
 *
 *
 *
 * Constraints:
 *
 *
 * 3 <= nums.length <= 3000
 * -10^5 <= nums[i] <= 10^5
 *
 *
 */

#include "leetcode.h"
// @lc code=start
class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    sort(nums.begin(), nums.end());
    auto ret = vector<vector<int>>();
    for (int i = 0; i < nums.size(); i++) {
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }
      int l = i + 1;
      int r = nums.size() - 1;
      int target = -1 * nums[i];
      while (l < r) {
        int sum = nums[l] + nums[r];
        if (sum == target) {
          ret.push_back(vector<int>({nums[i], nums[l], nums[r]}));
          l++;
          r--;
          while (l < r && nums[l] == nums[l - 1]) {
            l++;
          }
          while (l < r && nums[r] == nums[r + 1]) {
            r--;
          }
        } else if (sum > target) {
          r--;
        } else {
          l++;
        }
      }
    }
    return ret;
  }
};
// @lc code=end
